pkg/documentlog/impl.go

```diff
@@ -108,158 +86,132 @@ func (dl *documentLog) Subscribe(documentType string) DocumentQueue {
 }
 
 func (dl documentLog) GetDocumentContents(hash model.Hash) (io.ReadCloser, error) {
-	var entry *entry
-	dl.documentsMutex.ReadLock(func() {
-		entry = dl.documentHashIndex[hash.String()]
-	})
-	if entry == nil {
-		return nil, ErrUnknownDocument
-	}
-	// TODO: Nil/len check on Contents?
-	if entry.contents == nil {
+	contents, err := dl.store.ReadContents(hash)
+	if contents == nil {
 		return nil, ErrMissingDocumentContents
 	}
-	return NoopCloser{Reader: bytes.NewReader(*entry.contents)}, nil
+	return contents, err
 }
 
-func (dl *documentLog) HasDocument(hash model.Hash) bool {
-	var result bool
-	dl.documentsMutex.ReadLock(func() {
-		result = dl.documentHashIndex[hash.String()] != nil
-	})
-	return result
+func (dl *documentLog) AddMissingDocuments(documents []model.Document) ([]model.Hash, error) {
+	currentDocuments, err := dl.store.GetAll()
+	if err != nil {
+		return nil, err
+	}
+	missingContentHashes := make([]model.Hash, 0)
+	// This nested loop looks extremely inefficient but since both slices are sorted it should be relatively efficient in practice
+	for _, document := range documents {
+		exists := false
+		hasContents := false
+		for _, current := range currentDocuments {
+			if document.Hash.Equals(current.Hash) {
+				hasContents = current.HasContents
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			if err := dl.addDocument(document); err != nil {
+				return nil, err
+			}
+		}
+		if !hasContents {
+			missingContentHashes = append(missingContentHashes, document.Hash)
+		}
+	}
+	return missingContentHashes, nil
 }
 
-func (dl *documentLog) HasContentsForDocument(hash model.Hash) bool {
-	var result bool
-	dl.documentsMutex.ReadLock(func() {
-		result = dl.documentHashIndex[hash.String()] != nil && dl.documentHashIndex[hash.String()].contents != nil
-	})
-	return result
+func (dl *documentLog) HasContentsForDocument(hash model.Hash) (bool, error) {
+	document, err := dl.store.Get(hash)
+	if err != nil {
+		return false, err
+	}
+	if document == nil {
+		return false, nil
+	}
+	return document.HasContents, nil
 }
 
-func (dl *documentLog) AddDocumentWithContents(timestamp time.Time, documentType string, contents io.Reader) (model.Document, error) {
+func (dl *documentLog) AddDocumentWithContents(timestamp time.Time, documentType string, contents io.Reader) (*model.Document, error) {
 	buffer := new(bytes.Buffer)
 	if _, err := buffer.ReadFrom(contents); err != nil {
-		return model.Document{}, err
+		return nil, err
 	}
 	document := model.Document{
 		Hash:      model.CalculateDocumentHash(documentType, timestamp, buffer.Bytes()),
 		Type:      documentType,
 		Timestamp: timestamp,
 	}
-	if dl.HasContentsForDocument(document.Hash) {
-		return model.Document{}, fmt.Errorf("document already exists (with content) for hash: %s", document.Hash)
+	if err := dl.addDocument(document); err != nil {
+		return nil, err
 	}
-	dl.AddDocument(document)
 	return dl.AddDocumentContents(document.Hash, buffer)
```

Comment:
True. If the second one fails, the document (metadata) is added but contents are missing and an error is returned. When the function is called again, the document won't be added a second time (since it already exists) and its contents will be added.